### PR TITLE
Fixed #2729 conflicting getters/setters when not respecting capitaliz…

### LINKED ISSUE #2729: JsonMappingException: Conflicting getter / setter definitions when using property names that start with set, get or is
I started to experience the following issue starting from 2.10 in combination with the jackson scala module.

Consider a POJO like this:

```java
class SamplePojo {
    private String settlementDate;
    private String getaways;
    private Boolean island;

    public void settlementDate(String v) { throw new Error("Should not get called"); }
    public String settlementDate() { throw new Error("Should not get called"); }
    public void setSettlementDate(String v) { settlementDate = v; }
    public String getSettlementDate() { return settlementDate; }

    public void getaways(String v) { throw new Error("Should not get called"); }
    public String getaways() { throw new Error("Should not get called"); }
    public void setGetaways(String v) { getaways = v; }
    public String getGetaways() { return getaways; }

    public void island(Boolean v) { throw new Error("Should not get called"); }
    public Boolean island() { throw new Error("Should not get called"); }
    public void setIsland(Boolean v) { island = v; }
    public Boolean isIsland() { return island; }
  }
```

When serializing, one will receive a `JsonMappingException` because both `getaways()` and `getGetaways()`, as well as `island()` and `isIsland()` are considered conflicting getters with the same priority.

When deserializing, both `settlementDate(String)` and `setSettlementDate(String)` will be conflicting setters with the same priority.

Case-sensitivity should be respected when determining the priority of setters and getters, i.e. `settlementDate()` should not be the same as `setTlementDate()`.

Also see `_setterPriority()` in `POJOPropertyBuilder` (as well as `_getterPriority()`):
```java
    protected int _setterPriority(AnnotatedMethod m)
    {
        final String name = m.getName();
        if (name.startsWith("set") && name.length() > 3) {
            // should we check capitalization?
            return 1;
        }
        return 2;
    }
```

I'll submit a pull request to have this fixed later today.

### PR DESCRIPTION
**Pull request by janeisklar**
…ation

### CODE REVIEW

**Comment by cowtowncoder** [comment]
Makes sense, although needs to be re-based for 2.12 as it adds new feature (apologies for not mentioning this earlier). I could perhaps access this for 2.11(.1) alternatively, with caveat that new feature would be sort of "dark launch"ed -- available, but not yet documented.
But since 2.10 is quite far down patch versions I don't think it should go there.

Thank you especially for adding Scala-aspect in tests: I think this makes more sense now... I think am not sure if Scala module's logic is sound wrt how Scala creates ... case classes?
But I know there are challenges that Scala and Kotlin have with Jackson introspection and their added data/case classes in which definition starts with field name, and accessors are created from that; whereas Jackson starts from accessors trying to match field name. This has caused a few issues over time regarding casing.

One last thing: if I have not asked for and received CLA yet (apologies if I already have), we'd need it. It's here:

https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf

and usually print, fill & sign, scan/photo, email to `info` at fasterxml dot com is the easiest way to get it to us. Only needs to be done once before the first contribution, good for all Jackson projects.

Thank you again for very quick turnaround time; looking forward to merging this patch!

**Comment by janeisklar** [comment]
To hopefully make everything a bit clearer I added a [minimal project](https://github.com/janeisklar/jackson-databind-bug-reproducer) that reproduces the issue within the setup we are using.

When running the project with `clean verify --fail-at-end` you will see the following versions fail due to the altered behaviour:

> [INFO] dto ................................................ SUCCESS [  4.133 s]
> [INFO] jackson-10-9 ....................................... SUCCESS [  1.649 s]
> [INFO] jackson-10-10 ...................................... SUCCESS [  1.211 s]
> [INFO] jackson-10-11 ...................................... SUCCESS [  1.204 s]
> [INFO] jackson-10-9-with-scala ............................ SUCCESS [  1.777 s]
> [INFO] jackson-10-10-with-scala ........................... FAILURE [  1.774 s]
> [INFO] jackson-10-11-with-scala ........................... FAILURE [  1.960 s]

The relevant exception is this one:
> com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Conflicting setter definitions for property "settlementDate": test.model.SetInSetter#settlementDate(1 params) vs test.model.SetInSetter#setSettlementDate(1 params)
>  at [Source: (String)"{"settlementDate":"foo"}"; line: 1, column: 1]
> 
> 	at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:62)
> 	at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.buildBeanDeserializer(BeanDeserializerFactory.java:227)
> 	at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.createBeanDeserializer(BeanDeserializerFactory.java:143)
> 	at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer2(DeserializerCache.java:414)
> 	at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer(DeserializerCache.java:349)
> 	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:264)
> 	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244)
> 	at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142)
> 	at com.fasterxml.jackson.databind.DeserializationContext.findRootValueDeserializer(DeserializationContext.java:491)
> 	at com.fasterxml.jackson.databind.ObjectMapper._findRootDeserializer(ObjectMapper.java:4669)
> 	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4478)
> 	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3434)
> 	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3402)
> 	at test.SerializationTest.setInSetterTest(SerializationTest.java:33)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:686)
> 	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
> 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
> 	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
> 	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
> 	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
> 	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
> 	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
> 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
> 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
> 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
> 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
> 	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
> 	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
> 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:212)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:208)
> 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:137)
> 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:71)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:135)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
> 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
> 	at java.util.ArrayList.forEach(ArrayList.java:1257)
> 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
> 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
> 	at java.util.ArrayList.forEach(ArrayList.java:1257)
> 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
> 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
> 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
> 	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
> 	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
> 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:248)
> 	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$5(DefaultLauncher.java:211)
> 	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:226)
> 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:199)
> 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:132)
> 	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:69)
> 	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
> 	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
> 	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)

If I changed the pull request such that this becomes an optional flag that is deactivated by default, would that be an acceptable approach to move this forward?

File: src/main/java/com/fasterxml/jackson/databind/introspect/POJOPropertyBuilder.java [hunk 1]
```diff
@@ -593,13 +592,19 @@ protected int _getterPriority(AnnotatedMethod m)
...
             return 1;
         }
         return 2;
     }
 
+    private boolean validateAccessorPrefix(String name, String prefix)
+    {
+        final int prefixLength = prefix.length();
+        return name.startsWith(prefix) && name.length() > prefixLength
+            && (!_config.isEnabled(MapperFeature.REQUIRE_CAPITALIZED_PROPERTY_ACCESSOR_NAME) || Character.isUpperCase(name.charAt(prefixLength)));
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/introspect/POJOPropertyBuilder.java:605]: One question here would be whether upper-case is required or lower-case letter not allowed -- difference being between "island" and "is_land()". I realize neither is Beans naming compatible, but I suspect latter might also be in use.

**Comment by cowtowncoder** [comment]
While I think it's a good idea to offer configurability, I'll close this one since there is #2800, and since it may be necessary to think of best way to offer easy configurability (there are some challenges wrt `MapperFeature` right now).

**Pull request closed without merging**